pilot/pkg/networking/plugin/httpmx/httpmx.go

```diff
@@ -0,0 +1,109 @@
+package httpmx
+
+import (
+	v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	wasm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/wasm/v3"
+	http_conn "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	wasm_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/wasm/v3"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking"
+	"istio.io/istio/pilot/pkg/networking/plugin"
+	"istio.io/istio/pilot/pkg/networking/util"
+)
+
+// Plugin implements Istio Telemetry HTTP metadata exchange
+type Plugin struct{}
+
+// NewPlugin returns an instance of the httpmx plugin
+func NewPlugin() plugin.Plugin {
+	return Plugin{}
+}
+
+// OnInboundListener is called whenever a new listener is added to the LDS output for a given service
+// Can be used to add additional filters.
+func (p Plugin) OnInboundListener(in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	if in.Node.Type != model.SidecarProxy {
+		// Only care about sidecar.
+		return nil
+	}
+	return buildFilter(in, mutable, false)
+}
+
+// OnInboundFilterChains is called whenever a plugin needs to setup the filter chains, including relevant filter chain
+// configuration, like FilterChainMatch and TLSContext.
+func (p Plugin) OnInboundFilterChains(in *plugin.InputParams) []networking.FilterChain {
+	httpmxFilter := buildHTTPMxFilter()
+	fc := networking.FilterChain{HTTP: []*http_conn.HttpFilter{httpmxFilter}}
+	return []networking.FilterChain{fc}
+}
+
+// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service.
+// Can be used to add additional filters on the outbound path.
+func (p Plugin) OnOutboundListener(in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	return nil
+}
+
+// OnInboundPassthrough is called whenever a new passthrough filter chain is added to the LDS output.
+// Can be used to add additional filters.
+func (p Plugin) OnInboundPassthrough(in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	return nil
+}
+
+// OnInboundPassthroughFilterChains is called whenever a plugin needs to setup custom pass through filter chain.
+func (p Plugin) OnInboundPassthroughFilterChains(in *plugin.InputParams) []networking.FilterChain {
+	return nil
+}
+
+func buildFilter(in *plugin.InputParams, mutable *networking.MutableObjects, isPassthrough bool) error {
+	for i := range mutable.FilterChains {
+		// todo: if wasm enabled or not?
+		// switch tpc protocol?
+		if in.ListenerProtocol == networking.ListenerProtocolHTTP || mutable.FilterChains[i].ListenerProtocol == networking.ListenerProtocolHTTP {
+			// if wasm not enabled: build pure http mx
+			if httpMxFilter := buildHTTPMxFilter(); httpMxFilter != nil {
+				mutable.FilterChains[i].HTTP = append(mutable.FilterChains[i].HTTP, httpMxFilter)
+			}
+		}
+	}
+
+	return nil
+}
+
+func buildHTTPMxFilter() *http_conn.HttpFilter {
```

Comment:
Can we make this a const

Reply:
Yes, can be, looking for a better place to organize this, John has suggested to add it in the xds filters.go file, I'm checking about it.